File: go/merge.go
package skg

// MergeNodes overlays `overlay` on top of `base`, returning a new slice.
//
// Fields with the same key: overlay wins (last-wins).
// Blocks with the same name: children merged recursively.
// New keys/blocks from overlay are appended.
func MergeNodes(base, overlay []Node) []Node {
	if len(overlay) == 0 {
		return base
	}

	result := make([]Node, 0, len(base)+len(overlay))
	index := make(map[string]int, len(base)+len(overlay))

	for _, n := range base {
		key := nodeKey(n)
		index[key] = len(result)
		result = append(result, n)
	}

	for _, ov := range overlay {
		key := nodeKey(ov)
		if pos, ok := index[key]; ok {
			if ov.Block != nil && result[pos].Block != nil {
				merged := MergeNodes(result[pos].Block.Children, ov.Block.Children)
				result[pos] = Node{Block: &Block{
					Name:     result[pos].Block.Name,
					Children: merged,
					Line:     result[pos].Block.Line,
					Col:      result[pos].Block.Col,
				}}
			} else {
				result[pos] = ov
			}
		} else {
			index[key] = len(result)
			result = append(result, ov)
		}
	}

	return result
}

func nodeKey(n Node) string {
	if n.Field != nil {
		return n.Field.Key
	}
	return n.Block.Name
}

File: zig/merge.zig
/// SKG merge - overlays one node list on top of another.
///
/// Used by root.zig to apply import files before the main file.
/// Load order rule: later values overwrite earlier values.
/// Block children are merged recursively by name.
///
/// All allocations go into `allocator`. With an arena, this is free to discard.
const std = @import("std");
const Allocator = std.mem.Allocator;
const ast = @import("ast.zig");

/// Merge `overlay` nodes on top of `base`, returning a new owned slice.
///
/// - Fields with the same key: overlay wins.
/// - Blocks with the same name: children merged recursively, overlay children win.
/// - New keys/blocks from overlay are appended.
///
/// Neither `base` nor `overlay` is modified. The result may reference elements
/// from both - only free via the allocator/arena that owns this memory.
pub fn mergeNodes(allocator: Allocator, base: []const ast.Node, overlay: []const ast.Node) ![]ast.Node {
    var result: std.ArrayListUnmanaged(ast.Node) = .empty;
    var index = std.StringHashMapUnmanaged(usize){};

    for (base) |node| {
        const key = nodeKey(node);
        const pos = result.items.len;
        try result.append(allocator, node);
        try index.put(allocator, key, pos);
    }

    for (overlay) |ov_node| {
        const key = nodeKey(ov_node);
        if (index.get(key)) |pos| {
            switch (ov_node) {
                .field => result.items[pos] = ov_node,
                .block => |ov_block| {
                    const existing = result.items[pos].block;
                    result.items[pos] = ast.Node{ .block = .{
                        .name = existing.name,
                        .children = try mergeNodes(allocator, existing.children, ov_block.children),
                        .line = existing.line,
                        .col = existing.col,
                    } };
                },
            }
        } else {
            const pos = result.items.len;
            try result.append(allocator, ov_node);
            try index.put(allocator, key, pos);
        }
    }

    return result.toOwnedSlice(allocator);
}

fn nodeKey(node: ast.Node) []const u8 {
    return switch (node) {
        .field => |f| f.key,
        .block => |b| b.name,
    };
}
